[Step20] log in and log out

snowmint opened the pull request:
步驟20: 登入/登出功能
這裡不使用任何 Gem，請自己實做
不使用 devise 等便利的 Gem，是為了讓新人能更深入了解 Rails 中 HTTP cookie 和 session 的原理
以及加強對一般認證機制的理解（例如密碼的處理）
- 實做登入的頁面
- 未登入時，不能進入任務管理頁面
- 請改成只能看到自己建立的任務
- 實做登出功能

snowmint committed:
[Step 19/ 20 in Progress] Add user and signup

snowmint committed:
[Step20] log in and log out

snowmint committed:
[Fix] UsersController rather than UserController

ryudoawaru commented on app/controllers/tasks_controller.rb:
```diff
@@ -1,85 +1,87 @@
+  def index
+    @tasks = Task.filtered(query_params).page(params[:page]).per(5)
+    @tasks = case params[:order]
+      when 'by_sort_create_time_desc'
```
請活用 String Class 提供的方式或是 RegExp 去拆解字串達成簡潔拆分 sort column & direction 的目的，提示：

[String#split](https://ruby-doc.org/core-2.6/String.html#method-i-split) + 以空白分隔的 `params[:order]`


或是

最低限度用 Hash 的方式索引

ex: 

```ruby
SORT_DIRECTION_REFLECTION = {
  by_sort_create_time_asc: 'created_at ASC',
  ... 以下省略
}
```

但盡量希望用第一種方式完成

ryudoawaru commented on app/models/user.rb:
```diff
@@ -6,17 +6,29 @@ class User < ApplicationRecord
+
   #method
+  def authenticate(login_params)
+    auth_user = User.find_by("username = ? AND password = ?", login_params[:username], login_params[:password]) # login, password are your users table fields…
```
這邊密碼應該是沒加密的？

雖然作業說明上沒明寫，但一般來說是不允許明碼存在的，請參考 [has_secure_password](https://api.rubyonrails.org/classes/ActiveModel/SecurePassword/ClassMethods.html)

或是自己用 MD5 之類的內建 class 達成存儲非明碼密碼+ salt 的目的 

另外也需要做到在註冊時能有 password confirmation 確認輸入兩次密碼做比對的機制

ryudoawaru commented:
請處理完現在的 comment 再通知我，謝謝！

snowmint committed:
[Modify] sort column & direction by string split

snowmint committed:
[Modify] Use has_secure_password

snowmint commented:
使用 string split 方式排序︰
```
@tasks = Task.by_user_id(session[:user_id]).filtered(query_params).page(params[:page]).per(5).order(sort_column + " " + sort_direction)
```

使用 has_secure_password 達到加密，原先套件中下載到的 'bcrypt' 版本為 '3.1.7' 一直遇到 Invaild hash 報錯，未改變任何程式的情況下升級至 '3.1.15' 便可正常使用

ryudoawaru commented:
可以再 review 時跟我說

snowmint commented:
> 可以再 review 時跟我說

前面提到的兩個問題都已經解決了，可能需要麻煩你幫我 review 一下，謝謝

ryudoawaru commented on app/controllers/application_controller.rb:
```diff
@@ -9,4 +9,22 @@ def set_locale
+      if params[:order]
+        Task.column_names.include?(params[:order].split()[0]) ? params[:order].split()[0] : "id"
+      else
+        Task.column_names.include?("id") ? "id" : "id"
```
應該不需要這樣

id 是一定會有的欄位

ryudoawaru commented on app/controllers/application_controller.rb:
```diff
@@ -9,4 +9,22 @@ def set_locale
+
+    def sort_column
+      if params[:order]
+        Task.column_names.include?(params[:order].split()[0]) ? params[:order].split()[0] : "id"
```
應該先用一個變數存 split 的結果

否則這樣你會執行 split 兩次

ryudoawaru commented on app/models/user.rb:
```diff
@@ -4,19 +4,34 @@ class User < ApplicationRecord
   }
+
   #method
+  # def authenticate(login_params)
```
不用的 code 應該在提交 PR 前刪除

除非這個之後還會用到

ryudoawaru commented on app/controllers/application_controller.rb:
```diff
@@ -9,4 +9,22 @@ def set_locale
+      end
+    end
+    
+    include SessionsHelper
```
include 應該放在頭

snowmint committed:
[Modify/Step21 prepare] modify and add try to add admin page

snowmint commented:
修正 Comment 提到的錯誤
新增 Step21 的管理者登入，尚未完成所有功能

ryudoawaru commented on app/controllers/application_controller.rb:
```diff
@@ -1,12 +1,31 @@
+  
+  def sort_column
+    if params[:order]
+      split_order = params[:order].split()
```
ruby 一般不用特別加 `()` 無參數呼叫 method

（這不用特別改）

ryudoawaru commented on app/controllers/application_controller.rb:
```diff
@@ -1,12 +1,31 @@
+    if params[:order]
+      split_order = params[:order].split()
+      Task.column_names.include?(split_order[0]) ? split_order[0] : "id"
+    else
```
其實這一層 if else 應該是不需要的
`params[:order].to_s.split` 的話 即使 `params[:order]` 是 nil 也會出來空字串而已

你可以用 rails  console  下去模擬這個狀況或是用 byebug 測試

snowmint replied:
好的，已修改

ryudoawaru commented on app/controllers/application_controller.rb:
```diff
@@ -1,12 +1,31 @@
+  def sort_direction
+    if params[:order]
+      split_order = params[:order].split()
+      %w[asc desc].include?(split_order[1]) ? split_order[1] : "asc"
```
同上

snowmint committed:
[Modify] remove the if condition that useless

ryudoawaru commented on app/helpers/users_helper.rb:
```diff
@@ -0,0 +1,9 @@
+    if x == true
+      x = I18n.t('user_relate.role_admin')
+    else
+      x = I18n.t('user_relate.role_member')
```
```ruby
key = x ? 'admin' : 'member'
I18n.t(key, scope: :user_relate)
```

基本上 ruby 不需要 x == true 只要 if x 就可以

ryudoawaru commented on app/models/task.rb:
```diff
@@ -38,8 +38,16 @@ def self.by_keyword(keyword)
   end
 
+  def self.by_user_id(user_id)
+    if user_id.present?
```
present 要少用 因為有點耗資源
一般狀況下用 if user_id 或 if user_id.blank? 就好

snowmint replied:
好的，已經修正了

snowmint committed:
[Modify] Syantax issue

ryudoawaru commented on app/controllers/tasks_controller.rb:
```diff
@@ -1,85 +1,56 @@
-        when 'by_sort_priority_asc'
-          @tasks.order('priority ASC')
+  def index
+    @tasks = Task.by_user_id(session[:user_id]).filtered(query_params).page(params[:page]).per(5).order(sort_column + " " + sort_direction)
```
這邊有可能是沒 user 的狀況下找 task 的嗎？看起來應該是

如果是這樣，那也不適合用 session 直接當 key，這邊是有安全性的疑慮，例如 session 被放了不正確的 user_id 之類；我們應該要定義一個基礎的 scope method

```ruby
def current_user
  User.find(session[:user_id])
end

def current_scope
  current_user ? current_user.tasks : Task
end

def index
  @tasks = current_scope.....#接後面的 scopes
end

```

current_user 通常是跨 controller 共用的
我們在寫 Ruby 的東西時會希望減少每個 method 的複雜度，盡量去用多個 method 疊起來

ryudoawaru replied:
這邊還有一個理由

就是

你怎麼確定 edit update destroy 時的那筆 task 一定屬於這個 user?

ryudoawaru commented on app/controllers/tasks_controller.rb:
```diff
@@ -1,85 +1,56 @@
-      params.require(:task).permit(:user_id, :task_name, :content, :start_time, :end_time, :priority, :status)
-    end
+  def edit
+    @task = Task.find_by(id: params[:id])
```
請用 before_action 減少重複

這件應該不是第一次提到了

snowmint committed:
[Modify & Step21] use before_action to reduse repeat code / admin can delete the user

ryudoawaru commented on app/controllers/tasks_controller.rb:
```diff
@@ -1,7 +1,12 @@
 class TasksController < ApplicationController
+  before_action :task_find_by_id, only:[:edit, :update, :destroy]
+  
+  def current_scope
```
這種通常都是放 private

可以理解原因嗎？（不一定要在這次改 code，但我想聽你的回答）

snowmint replied:
是因為要避免別人直接透過 function 存取資料

ryudoawaru replied:
一般來說 controller public method 只會放 action，其它的都要 private

這樣子設計也會比較好區分哪個 method 是 action 與否

snowmint replied:
好的了解，跟新 push 上去的部份一起修改掉了

ryudoawaru commented on app/controllers/tasks_controller.rb:
```diff
@@ -52,5 +53,9 @@ def query_params
   end
+
+  def task_find_by_id
+    @task = Task.find(params['id'])
```
我前面有提過

你要怎麼確保我不會改到別人的 task?

ryudoawaru replied:
不好意思我發現有可能是你還在改，等你改完好了

snowmint committed:
[Modify & Step21] public part just can contain action / admin can delete the user

[Modify]
The public field can only contain action.

[Step21]
List the users and the number of tasks of each user.
The admin can delete a user, with their tasks.
Try to use admin permission to edit the user's data, with authentication.(not yet complete)

snowmint committed:
[Fix] broken

ryudoawaru commented:
之後要試著讓 PR 不要太大包，一般來說是 < 20 個檔案，理想是 < 15

可以把 1 step 分成幾個 PR 也 OK

snowmint commented:
> 之後要試著讓 PR 不要太大包，一般來說是 < 20 個檔案，理想是 < 15
> 
> 可以把 1 step 分成幾個 PR 也 OK

好的我會注意

ryudoawaru commented on app/controllers/tasks_controller.rb:
```diff
@@ -1,85 +1,66 @@
+
+  private
+
+    before_action :task_find_by_id, only:[:edit, :update, :destroy]
```
before action 使用的 method 要放在後面 private 區沒錯，但相反的 before_action 的設定要放在頂部，教材裡或是 Rails Guide 的範例應該都是這樣示範，原因是因為要增進 code 的易讀性，所以想像當有別人要改你的程式或是 review 時：

1. 從頭看到 before_action 的設定，可以知道大概有哪些東西
2. public 區放 actions
2. method 的命名需要「達意」，望名即知用途

如果團隊中乃至大部份的 Rails 開發者都是遵照這些 convention，那長久下來大家就不需要花很多時間適應彼此的習慣，共同開發時也會比較有效率。

snowmint replied:
那我可能誤解上面 comment 的意思了，我會將 before_action 放回頂部

ryudoawaru commented on app/controllers/application_controller.rb:
```diff
@@ -1,12 +1,23 @@
-      I18n.locale = session[:locale] || I18n.default_locale
+  def set_locale
+    # 可以將 ["en", "zh-TW"] 設定為 VALID_LANG 放到 config/environment.rb 中
+    if params[:locale] && I18n.available_locales.include?( params[:locale].to_sym )
```
（不用特別改的部份）

只是要知道有沒有 key，其實用 `Hash#has_key?` 就好了，另外以這邊 include? 的寫法可以再更簡潔

`I18n.available_locales.include?(params.dig(:locale).to_s.to_sym)` 即可

我們會盡量活用 Ruby 的萬物皆物件與 chain method 的特性

ryudoawaru commented on app/controllers/application_controller.rb:
```diff
@@ -1,12 +1,23 @@
+  def set_locale
+    # 可以將 ["en", "zh-TW"] 設定為 VALID_LANG 放到 config/environment.rb 中
+    if params[:locale] && I18n.available_locales.include?( params[:locale].to_sym )
+      session[:locale] = params[:locale]
```
然後以這種 oo = xx if YY 的 case 我們都用用 guard clause 也就是一行去完成

```ruby
session[:locale] = params[:locale] If I18n.available_locales.include?(params[:locale]).to_s.to_sym)
```

ryudoawaru commented on app/controllers/tasks_controller.rb:
```diff
@@ -1,85 +1,66 @@
-  end
+
+    def task_find_by_id
+      @tasks = Task.find(params['id'])
```
這邊你還是沒有用到 current_scope

而且這邊應該多了一個 s

snowmint replied:
在 index 中有使用到 current_scope，不知道正不正確？
而 s 的部份是指 current_user.task"s" 嗎？這裡的話有用 console 試過需要加 s 才能透過 user 取得對應的 tasks 資料

ryudoawaru replied:
不是 是你這行
`@tasks = Task.find(params['id'])` 是 `@tasks`

但這是單數的東西

而且後面 edit/update/delete 用的時候你 action 裡寫的應該都是 `@task` 沒 s

然後我前面一直提示的答案，其實是這邊需要用 `current_scope.find` 而不是 `Task.find` 你可以試著說明為何是這樣嗎？

snowmint replied:
是因為 current_scope 是以 user 來找關聯的 tasks ，而不是在全部的 tasks 裡找，是效能上會比較有效率嗎？而且即使 task id 被修改 user 也只能找到屬於自己的 task 而已，看不到其他人的。

ryudoawaru replied:
重要的是後者

ryudoawaru commented on app/helpers/sessions_helper.rb:
```diff
@@ -0,0 +1,24 @@
+
+  def remember(user)
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
```
以作業的 spec 來說應該沒有要求要特地設 cookie?

ryudoawaru commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,70 @@
+
+    def correct_user
+      @user = User.find(params[:id])
+      redirect_to(root_path) unless @user == current_user || current_user.admin?
```
有 or 的狀況下要避免用 unless

比較容易讓看的人混淆

ryudoawaru commented on app/models/task.rb:
```diff
@@ -15,31 +15,39 @@ class Task < ApplicationRecord
   #class method
   def self.by_status(status)
-    if status.present?
+    if !status.blank?
```
if 又 !  會降低易讀性，這樣 unless 就好，而且在你有 else 的情況下順序對換就好

ryudoawaru replied:
另外這邊也是適用三元運算子的 case

snowmint replied:
改成三元運算子的方式了

ryudoawaru commented:
請照 review 繼續修改

snowmint committed:
[Modify] With the comment

snowmint committed:
[Modify] current scope

snowmint committed:
[Modify] current scope

snowmint commented:
上面提到的問題已經修正，可能需要麻煩 review 一下

暫時開了另一個 branch 避免 push 時的程式跟這裡修改的部份混在一起，如果 review 完這裡我會刪掉暫時的 branch ，再把新加上的部份 push 到 topic branch

ryudoawaru commented on app/controllers/application_controller.rb:
```diff
@@ -1,12 +1,19 @@
+  end
+  
+  def sort_column
+    split_order = params[:order].to_s.split()
```
1. method 呼叫沒參數就不要加 ()
2. split_order 在兩個 method 中重複，應該要抽離成另一個 method 並且加上 instance variable 以減少 cost

```ruby

def split_order_cmds
  @split_order_cmds ||= params[:order].to_s.split
end

```

後面不管呼叫幾次都會返回 instance var

ryudoawaru commented:
下次改好了請早點說

snowmint committed:
[Modify] split cmds

snowmint commented:
> 下次改好了請早點說

好的，之後會直接開一個新的 comment 來說，而不是只有 reply comment
split_order 部份也改好了
